daemon/src/wire.rs

```diff
@@ -80,7 +80,9 @@ pub mod taker_to_maker {
 #[serde(tag = "type", content = "payload")]
 #[allow(clippy::large_enum_variant)]
 pub enum TakerToMaker {
-    Hello(Version),
+    Hello {
+        wire_version: Version,
```

Comment:
I am not sure we should break _this_ message as well. With a blue/green deployment, we should be fine doing this but for older clients that try to connect to this version, the version negotiation _itself_ will break.

Normally that is a no-no in terms of compatibility. You shouldn't break the message that establishes whether or not you are compatible.

Reply:
Fair enough, I have relied implicitly on the blue/green deployment strategy we're about to use to get over the big protocol break, and that could have cause short-term breakage if it wasn't used :/

I could add a whole additional message, but this will complicate the protocol and bake in a temporary awkwardness forever.

What would you consider to be the best path forward?

Reply:
If we wanted to be backwards-compatible, we can just make the `daemon_version` field optional. I would do that like this:

1. Write a test that cements the serialization of the current `Hello` message
2. Introduce a newtype for the payload of the `Hello` message
3. Add an optional, 2nd field to the payload struct and write more tests against it to ensure we can still deserialize it correctly

Probably need to make use of `serde(default)` or something like that.

Reply:
If you wanted to introduce the named fields as well, I you could always write a custom deserializer for the payload struct as well.

Reply:
Probably can actually skip the newtype if you don't want a custom serializer.

Reply:
Thank you for these comments, they are really valuable if we care about it in this case.

Before I go off on this path, I'd like to confirm - do we actually want/need to be backwards compatible in this case, given we're doing the blue-green deployment this time?

Reply:
I'd say give how simple it seems to be, I'd try and implement this in a time-box. If you can't achieve it, then just break the API :)

Reply:
> Before I go off on this path, I'd like to confirm - do we actually want/need to be backwards compatible in this case, given we're doing the blue-green deployment this time?

I don't think it matters at this point, given the changes when receiving `Offers`. However, as Thomas pointed out, the user will not get an error messages that can be interpreted if we break `Hello`. However, I don't think many users look at the logs. I'm OK with breaking changes to `Hello` at this point, but agree with Thomas that we can try to do this different if easily possible.

Reply:
Timeboxing myself on doing this now, will try to address it before midday :-)

Reply:
Didn't manage to finish it within the timebox - as spoken with @thomaseizinger, the message will get broken soon anyway with the introduction of libp2p, so I don't think it's worth investing more time into it nor delaying the release.

Reply:
update: I tried using advanced features of serde, but making it work in the way outlined above turned out to be impossible as serde would would serialise things as an array (`[]`) the moment we add a second field, even into a tuple struct or inside a newtype :(